### PR TITLE
New Adapter: IQZone

### PR DESCRIPTION
**Pull request by IQZoneAdx**
Doc - https://github.com/prebid/prebid.github.io/pull/3210

### CODE REVIEW

File: openrtb_ext/imp_iqzone.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package openrtb_ext
+
+type ImpExtFoo struct {
```
Comment (guscarreon) [inline openrtb_ext/imp_iqzone.go:3]: Even if it's not being used, please rename to something less generic, `ImpExtIqzone` maybe

File: static/bidder-params/iqzone.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    "type": "object",
+  
+    "properties": {
+      "placementId": {
+        "type": "string",
+        "description": "Placement ID"
+      }
+    },
+  
+    "required": ["placementId"]
```
Comment (guscarreon) [inline static/bidder-params/iqzone.json:14]: Just to double-check: Are we sure every request will come with a `"placementId"` parameter? If it's set as `"required"`, an incoming request with a missing `"placementId"` will get rejected and Prebid Server won't send a bid request to any bidder. So adding a required field can break the entire setup of a publisher.

File: adapters/iqzone/params_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		}
+	}
+}
+
+var validParams = []string{
+	`{"placementId": "test"}`,
+}
+
+var invalidParams = []string{
+	`{"placementId": 42}`,
```
Comment (guscarreon) [inline adapters/iqzone/params_test.go:41]: If you decide to keep the `"placementId"` listed in the `"required"` array inside `static/bidder-params/iqzone.json`, can we please add a couple more test cases in order to assert the request will be invalid when `"placementId"` isn't found in the `Ext`?
```
40   var invalidParams = []string{
41       `{"placementId": 42}`,
   +     `{}`,
   +     `{"someOtherParam": "value"}`,
42   }
adapters/iqzone/params_test.go
```

File: adapters/iqzone/iqzone.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bid := bid // pin https://github.com/kyoh86/scopelint#whats-this
```
Comment (guscarreon) [inline adapters/iqzone/iqzone.go:69]: Good catch, we can remove the comment if you want

File: adapters/iqzone/iqzone.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			if imp.Banner == nil && imp.Video == nil && imp.Native != nil {
+				mediaType = openrtb_ext.BidTypeNative
+			}
```
Comment (guscarreon) [inline adapters/iqzone/iqzone.go:89]: Is the `imp.Banner == nil && imp.Video == nil && imp.Native == nil` scenario possible? If so, should `getMediaTypeForImp` return an error instead?

**Review by guscarreon — CHANGES_REQUESTED** [review]
Left a few comments. Also, there is already a pull request to remove the old race condition test framework and an additional PR is under development to update the documents. We apologize for not having the documentation up to date yet. In the meantime, can we please remove the following files?
```
iqzone/iqzonetest/params/race/banner.json
iqzone/iqzonetest/params/race/native.json
iqzone/iqzonetest/params/race/video.json
```
These were required under the old data race tests and are no longer necessary. In regards of test case coverage your adapter looks file. It only seems to be missing the following scenarios inside `MakeBids`:
image
Can we please add a couple more JSON tests under `iqzone/iqzonetest/supplemental/` to cover them?

File: openrtb_ext/imp_iqzone.go (hunk 1 above)
Comment (SyntaxNode) [reply openrtb_ext/imp_iqzone.go:3]: This just looks like a copy/paste error from the docs. Easy fix. Thank you @IQZoneAdx for faithfully following our new bid adapter guide. :)

File: adapters/iqzone/iqzone.go (hunk 4 above)
Comment (SyntaxNode) [reply adapters/iqzone/iqzone.go:69]: This is a copy/paste from the docs. A clearer way of writing this loop could be:
```
for i := range seatBid.Bid {
  b := &adapters.TypedBid{
    Bid:     &seatBid.Bid[i],
    BidType: getMediaTypeForImp(bid.ImpID, request.Imp),
  }
  bidResponse.Bids = append(bidResponse.Bids, b)
}
```
This does look better IMHO and doesn't require extra explanation via the pin link. I'll update the docs soon with this approach.

**Comment by IQZoneAdx** [comment]
@guscarreon @SyntaxNode thanks for the review. We fixed all the problems

**Review by guscarreon — APPROVED** [review]
Looks good to me. @IQZoneAdx thank you for addressing the feedback and following the documentation steps so thoroughly.

**Review by SyntaxNode — APPROVED** [review]

**Pull request merged**